drivers/spi/spi_b_renesas_ra8.c

```diff
@@ -0,0 +1,771 @@
+{
+	R_SPI_B0_Type *p_spi_reg = data->spi.p_regs;
+
+	if (p_spi_reg->SPSR_b.SPTEF && spi_context_tx_on(&data->ctx)) {
+		uint32_t tx;
+
+		if (data->ctx.tx_buf != NULL) {
+			if (data->dfs > 2) {
+				tx = *(uint32_t *)(data->ctx.tx_buf);
+			} else if (data->dfs > 1) {
+				tx = *(uint16_t *)(data->ctx.tx_buf);
+			} else {
+				tx = *(uint8_t *)(data->ctx.tx_buf);
+			}
+		} else {
+			tx = 0;
+		}
+		/* Clear Transmit Empty flag */
+		p_spi_reg->SPSRC = R_SPI_B0_SPSRC_SPTEFC_Msk;
+
+		p_spi_reg->SPDR = tx;
+
+		spi_context_update_tx(&data->ctx, data->dfs, 1);
+	} else {
+		p_spi_reg->SPCR_b.SPTIE = 0;
+	}
+
+	if (p_spi_reg->SPSR_b.SPRF && spi_context_rx_buf_on(&data->ctx)) {
+		uint32_t rx;
+
+		rx = p_spi_reg->SPDR;
+		/* Clear Receive Full flag */
+		p_spi_reg->SPSRC = R_SPI_B0_SPSRC_SPRFC_Msk;
+		if (data->dfs > 2) {
+			UNALIGNED_PUT(rx, (uint32_t *)data->ctx.rx_buf);
+		} else if (data->dfs > 1) {
+			UNALIGNED_PUT(rx, (uint16_t *)data->ctx.rx_buf);
+		} else {
+			UNALIGNED_PUT(rx, (uint8_t *)data->ctx.rx_buf);
+		}
+		spi_context_update_rx(&data->ctx, data->dfs, 1);
+	}
+
+	return 0;
+}
+
+static int ra_spi_b_transceive_master(struct ra_spi_data *data)
+{
+	R_SPI_B0_Type *p_spi_reg = data->spi.p_regs;
+	uint32_t tx;
+	uint32_t rx;
+
+	/* Tx transfer*/
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		if (data->dfs > 2) {
+			tx = *(uint32_t *)(data->ctx.tx_buf);
+		} else if (data->dfs > 1) {
+			tx = *(uint16_t *)(data->ctx.tx_buf);
+		} else {
+			tx = *(uint8_t *)(data->ctx.tx_buf);
+		}
+	} else {
+		tx = 0U;
+	}
+
+	while (!p_spi_reg->SPSR_b.SPTEF) {
+	}
+
+	p_spi_reg->SPDR = tx;
+
+	/* Clear Transmit Empty flag */
+	p_spi_reg->SPSRC = R_SPI_B0_SPSRC_SPTEFC_Msk;
+	spi_context_update_tx(&data->ctx, data->dfs, 1);
+
+	/* Rx receive */
+	if (spi_context_rx_on(&data->ctx)) {
+		while (!p_spi_reg->SPSR_b.SPRF) {
+		}
+		rx = p_spi_reg->SPDR;
+		/* Clear Receive Full flag */
+		p_spi_reg->SPSRC = R_SPI_B0_SPSRC_SPRFC_Msk;
+		if (data->dfs > 2) {
+			UNALIGNED_PUT(rx, (uint32_t *)data->ctx.rx_buf);
+		} else if (data->dfs > 1) {
+			UNALIGNED_PUT(rx, (uint16_t *)data->ctx.rx_buf);
+		} else {
+			UNALIGNED_PUT(rx, (uint8_t *)data->ctx.rx_buf);
+		}
+		spi_context_update_rx(&data->ctx, data->dfs, 1);
+	}
+
+	return 0;
+}
+
+static int ra_spi_b_transceive_data(struct ra_spi_data *data)
+{
+	uint16_t operation = data->ctx.config->operation;
+
+	if (SPI_OP_MODE_GET(operation) == SPI_OP_MODE_MASTER) {
+		ra_spi_b_transceive_master(data);
+	} else {
+		ra_spi_b_transceive_slave(data);
+	}
+
+	return 0;
+}
+#endif
+
+static int transceive(const struct device *dev, const struct spi_config *config,
+		      const struct spi_buf_set *tx_bufs, const struct spi_buf_set *rx_bufs,
+		      bool asynchronous, spi_callback_t cb, void *userdata)
+{
+	struct ra_spi_data *data = dev->data;
+	R_SPI_B0_Type *p_spi_reg;
+	int ret = 0;
+
+	if (!tx_bufs && !rx_bufs) {
+		return 0;
+	}
+
+#ifndef CONFIG_SPI_B_INTERRUPT
```

Comment:
This is not blocking, but
async and interrupt-driven are essentially independent.
In other words, in the case of ASYNC, asynchrony is achieved using the argument k_poll_signal, so there is essentially no problem with a polling implementation that does not use interrupts.